imports/plugins/core/core/server/Reaction/addRolesToGroups.js

```diff
@@ -77,17 +29,76 @@ export default function addRolesToGroups(options = { allShops: false, roles: [],
       $in: groups
     }
   };
-  const multi = { multi: true };
 
   if (!allShops) {
-    // if we're not updating for all shops, we should only update for the shops passed in.
+    // If we're not updating for all shops, we should only update for the shops passed in.
     query.shopId = { $in: shops || [] };
     Logger.debug(`Adding Roles: ${roles} to Groups: ${groups} for shops: ${shops}`);
   } else {
     Logger.debug(`Adding Roles: ${roles} to Groups: ${groups} for all shops`);
   }
 
-  addRolesToUsersInGroups({ query, roles });
+  // Check if each group already has the roles. If so, skip updates for group & its users
+  Groups.find(query).forEach((group) => {
+    let areRolesInGroup = true;
+    roles.forEach((role) => {
+      if (group.permissions.includes(role) === false) {
+        Logger.debug(`Role ${role} is not in ${group.name} group`);
+        areRolesInGroup = false;
+      }
+    });
+
+    if (areRolesInGroup === false) {
+      Logger.debug(`Adding roles (${roles}) to ${group.name} group (${group._id})`);
+      // Run heaviest queries asynchronously in the background
+      Meteor.defer(() => {
+        addRolesToGroupAndUsers(group, roles);
+      });
+    } else {
+      Logger.debug(`Skipping roles already assigned to ${group.name} group and its users`);
+    }
+  });
+}
+
+/**
+ * @name addRolesToGroupAndUsers
+ * @private
+ * @summary Adds the given roles to the given group, and updates users if necessary
+ * @param {Object} group - Group to add roles to, and update users for
+ * @param {String} group._id - Group's _id
+ * @param {String} group.shopId - _id of shop group belongs to
+ * @param {String} group.name - Name of group
+ * @param {Array} roles - Array of roles/permissions (strings)
+ * @returns {undefined} Nothing
+ */
+function addRolesToGroupAndUsers({ _id, shopId, name }, roles) {
+  Groups.update({ _id }, { $addToSet: { permissions: { $each: roles } } });
+
+  const maxAccountsPerUpdate = 100000;
+  const accountSelector = { groups: _id };
+  const accountOptions = { fields: { _id: 1 } };
+  const numAccounts = Accounts.find(accountSelector, accountOptions).count();
+  if (numAccounts === 0) {
+    Logger.debug("No users need roles updated");
```

Comment:
I think there should be `return` statement here.

Reply:
@Akarshit Ah yep, you're right